seq2science/rules/trimming.smk

```diff
@@ -1,76 +1,142 @@
-        """\
-        trim_galore --paired -j {threads} {params.config} -o $(dirname {output.r1}) {input.r1} {input.r2} > {log} 2>&1
 
-        # now rename to proper output
-        for f in $(find "$(dirname {output.r1})/" -name "{wildcards.sample}_*val_*.fq.gz"); do
-            mv "$f" "$(echo "$f" | sed s/.fq/.{params.fqsuffix}/)"; done
-        for f in $(find "$(dirname {output.r1})/" -maxdepth 1 -name "{wildcards.sample}_*val_*.{params.fqsuffix}.gz"); do
-            mv "$f" "$(echo "$f" | sed s/_val_./_trimmed/)"; done
+    rule fastp_SE:
+        """
+        Automated adapter detection, adapter trimming, and quality trimming through fastp (single-end).
+        """
+        input:
+            expand("{fastq_dir}/{{sample}}.{fqsuffix}.gz", **config),
+        output:
+            se=temp(expand("{trimmed_dir}/{{sample}}_trimmed.{fqsuffix}.gz", **config)),
+            qc_json=expand("{qc_dir}/trimming/{{sample}}.fastp.json", **config),
+            qc_html=expand("{qc_dir}/trimming/{{sample}}.fastp.html", **config),
+        conda:
+            "../envs/fastp.yaml"
+        threads: 3
+        message: explain_rule("fastp_SE")
+        wildcard_constraints:
+            sample="|".join([sample if config["layout"][sample] == "SINGLE" else "$a" for sample in all_samples])
+        log:
+            expand("{log_dir}/fastp_SE/{{sample}}.log", **config),
+        benchmark:
+            expand("{benchmark_dir}/fastp_SE/{{sample}}.benchmark.txt", **config)[0]
+        params:
+            fqsuffix=config["fqsuffix"],
+        shell:
+            """\
+            fastp -w {threads} --in1 {input} --out1 {output.se} -h {output.qc_html} -j {output.qc_json} \
+            > {log} 2>&1
+            """
 
-        # move the trimming reports to qc directory
-        for f in $(find "$(dirname {output.r1})/" -name "{wildcards.sample}_*.{params.fqsuffix}.gz_trimming_report.txt"); do
-            mv "$f" "$(dirname {output.qc[0]})/$(basename $f)"; done
+
+    rule fastp_PE:
+        """
+        Automated adapter detection, adapter trimming, and quality trimming through fastp (paired-end).
         """
+        input:
+            r1=expand("{fastq_dir}/{{sample}}_{fqext1}.{fqsuffix}.gz", **config),
+            r2=expand("{fastq_dir}/{{sample}}_{fqext2}.{fqsuffix}.gz", **config),
+        output:
+            r1=temp(expand("{trimmed_dir}/{{sample}}_{fqext1}_trimmed.{fqsuffix}.gz", **config)),
+            r2=temp(expand("{trimmed_dir}/{{sample}}_{fqext2}_trimmed.{fqsuffix}.gz", **config)),
+            qc_json=expand("{qc_dir}/trimming/{{sample}}.fastp.json", **config),
+            qc_html=expand("{qc_dir}/trimming/{{sample}}.fastp.html", **config),
+        conda:
+            "../envs/fastp.yaml"
+        threads: 3
+        wildcard_constraints:
+            sample="|".join([sample if config["layout"][sample] == "PAIRED" else "$a" for sample in all_samples])
```

Comment:
what is `$a` here? An impossible input, a warning message defined elsewhere?

Reply:
Yes, $ means end of string, so $a means match a letter a after the end of a string. Which can never be matched. We need to add those, because if we only add possible matches, if the list comprehension is empty we end up with an empty string which can always be matched.

Reply:
btw we need this (and not for fastqc) because ruleorder can not resolve only qc for technical replicates and no trimming